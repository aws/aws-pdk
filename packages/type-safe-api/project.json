{
  "name": "@aws/type-safe-api",
  "root": "packages/type-safe-api",
  "targets": {
    "build": {
      "inputs": [
        "default",
        "^default",
        "!{projectRoot}/.jsii",
        "!{projectRoot}/lib/**/*",
        "!{projectRoot}/dist/**/*",
        "!{projectRoot}/coverage/**/*",
        "!{projectRoot}/test-reports/**/*",
        "!{projectRoot}/docs/api/**/*"
      ],
      "outputs": [
        "{projectRoot}/.jsii",
        "{projectRoot}/lib",
        "{projectRoot}/dist",
        "{projectRoot}/coverage",
        "{projectRoot}/test-reports",
        "{projectRoot}/docs/api"
      ],
      "dependsOn": [
        "^build"
      ],
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen build",
        "cwd": "packages/type-safe-api"
      }
    },
    "default": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen default",
        "cwd": "packages/type-safe-api"
      }
    },
    "pre-compile": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen pre-compile",
        "cwd": "packages/type-safe-api"
      }
    },
    "compile": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen compile",
        "cwd": "packages/type-safe-api"
      }
    },
    "post-compile": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen post-compile",
        "cwd": "packages/type-safe-api"
      }
    },
    "test": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen test",
        "cwd": "packages/type-safe-api"
      }
    },
    "package": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen package",
        "cwd": "packages/type-safe-api"
      }
    },
    "test:watch": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen test:watch",
        "cwd": "packages/type-safe-api"
      }
    },
    "watch": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen watch",
        "cwd": "packages/type-safe-api"
      }
    },
    "eslint": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen eslint",
        "cwd": "packages/type-safe-api"
      }
    },
    "compat": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen compat",
        "cwd": "packages/type-safe-api"
      }
    },
    "package-all": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen package-all",
        "cwd": "packages/type-safe-api"
      }
    },
    "package:java": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen package:java",
        "cwd": "packages/type-safe-api"
      }
    },
    "package:python": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen package:python",
        "cwd": "packages/type-safe-api"
      }
    },
    "eslint-staged": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen eslint-staged",
        "cwd": "packages/type-safe-api"
      }
    },
    "jest": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen jest",
        "cwd": "packages/type-safe-api"
      }
    },
    "docgen": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen docgen",
        "cwd": "packages/type-safe-api"
      }
    }
  },
  "implicitDependencies": [
    "smithy-async-transformer"
  ],
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \"pnpm exec projen\"."
}
