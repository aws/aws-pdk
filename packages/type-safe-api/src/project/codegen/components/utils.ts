/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import * as path from "path";
import * as readPkg from "read-pkg-up";
import { Language, Library } from "../../languages";
import { MockResponseDataGenerationOptions } from "../../types";

/**
 * Enum for generator directories for non-runtime generators
 */
export enum OtherGenerators {
  DOCS = "docs",
  // Infrastructure
  TYPESCRIPT_CDK_INFRASTRUCTURE = "typescript-cdk-infrastructure",
  PYTHON_CDK_INFRASTRUCTURE = "python-cdk-infrastructure",
  JAVA_CDK_INFRASTRUCTURE = "java-cdk-infrastructure",
  // Handlers
  TYPESCRIPT_LAMBDA_HANDLERS = "typescript-lambda-handlers",
  PYTHON_LAMBDA_HANDLERS = "python-lambda-handlers",
  JAVA_LAMBDA_HANDLERS = "java-lambda-handlers",
}

export enum TypeSafeApiScript {
  PARSE_OPENAPI_SPEC = "parse-openapi-spec",
  GENERATE = "generate",
  GENERATE_MOCK_DATA = "generate-mock-data",
  GENERATE_HTML_REDOC_DOCS = "generate-html-redoc-docs",
  CLEAN_OPENAPI_GENERATED_CODE = "clean-openapi-generated-code",
  COPY_GRADLE_WRAPPER = "copy-gradle-wrapper",
}

/**
 * Generator directory for openapi generation containing templates, config etc.
 */
export type GeneratorDirectory = Language | Library | OtherGenerators;

/**
 * Types of normalizers supported by openapi-generator
 * @see https://openapi-generator.tech/docs/customization/#openapi-normalizer
 */
export type OpenApiNormalizer = "KEEP_ONLY_FIRST_TAG_IN_OPERATION";

/**
 * Options for generating code or docs using OpenAPI Generator CLI
 */
export interface GenerationOptions {
  /**
   * The OpenAPI generator to use to generate the code/docs
   */
  readonly generator: string;
  /**
   * The directory to use for OpenAPI generation
   */
  readonly generatorDirectory: GeneratorDirectory;
  /**
   * The path of the OpenAPI spec to generate code from
   */
  readonly specPath: string;
  /**
   * Additional properties to pass to the generate cli
   */
  readonly additionalProperties?: {
    [key: string]: string;
  };
  /**
   * Supply the relative path from the code project root to the source code directory in which custom generated files
   * (eg. operation config) should be placed.
   */
  readonly srcDir?: string;
  /**
   * Normalizers to apply to the spec prior to generation, if any
   * @see https://openapi-generator.tech/docs/customization/#openapi-normalizer
   */
  readonly normalizers?: Partial<Record<OpenApiNormalizer, boolean>>;
  /**
   * Vendor extensions to add for code generation, allowing custom properties to be passed to the generator templates
   * Keys should begin with "x-"
   */
  readonly extraVendorExtensions?: Record<string, string | boolean>;
  /**
   * Generate alias as model.
   * Defines whether "primitive types" defined at the model/schema level will be wrapped in a model (ie maps/lists)
   * @see https://openapi-generator.tech/docs/globals/
   * @default true
   */
  readonly generateAliasAsModel?: boolean;
  /**
   * The path to the json smithy model file, if available
   */
  readonly smithyJsonPath?: string;
}

/**
 * Build a command for running a script from this project's bin
 */
export const buildTypeSafeApiExecCommand = (
  script: TypeSafeApiScript,
  args?: string
) => {
  const { packageJson } = readPkg.sync({
    cwd: path.resolve(__dirname),
  })!;
  return `npx --yes -p aws-pdk@${packageJson.version} ${script}${
    args ? ` ${args}` : ""
  }`;
};

const serializeProperties = (properties: { [key: string]: string }) =>
  Object.entries(properties)
    .map(([key, value]) => `${key}=${value}`)
    .join(",");

/**
 * Generate code or docs by invoking the root generate script
 */
export const buildInvokeOpenApiGeneratorCommandArgs = (
  options: GenerationOptions
): string => {
  const srcDir = options.srcDir ?? "src";
  const additionalProperties = options.additionalProperties
    ? ` --additional-properties "${serializeProperties(
        options.additionalProperties
      )}"`
    : "";

  const normalizers = options.normalizers
    ? ` --openapi-normalizer "${serializeProperties(
        Object.fromEntries(
          Object.entries(options.normalizers).map(([k, v]) => [k, `${v}`])
        )
      )}"`
    : "";

  const extensions = options.extraVendorExtensions
    ? ` --extra-vendor-extensions '${JSON.stringify(
        options.extraVendorExtensions
      )}'`
    : "";

  const generateAliasAsModel =
    options.generateAliasAsModel ?? true ? " --generate-alias-as-model" : "";

  const smithyJsonPath = options.smithyJsonPath
    ? ` --smithy-json-path ${options.smithyJsonPath}`
    : "";

  const specPath = options.specPath;
  const outputPath = ".";

  return `--generator ${options.generator} --spec-path ${specPath} --output-path ${outputPath} --generator-dir ${options.generatorDirectory} --src-dir ${srcDir}${smithyJsonPath}${additionalProperties}${normalizers}${extensions}${generateAliasAsModel}`;
};

/**
 * Builds a command to clean up files which were previously generated by openapi generator
 */
export const buildCleanOpenApiGeneratedCodeCommand = (): string => {
  return buildTypeSafeApiExecCommand(
    TypeSafeApiScript.CLEAN_OPENAPI_GENERATED_CODE,
    `--code-path .`
  );
};

/**
 * Options for generating mock data json files
 */
export interface MockDataGenerationOptions
  extends MockResponseDataGenerationOptions {
  /**
   * The path of the OpenAPI spec to generate data for
   */
  readonly specPath: string;
  /**
   * Output sub directory relative to the outdir in which to generate mock data
   * Mock data will be written to a directory named 'mocks' within the sub directory
   * @default .
   */
  readonly outputSubDir?: string;
}

/**
 * Invoke the mock data generator script
 */
export const buildInvokeMockDataGeneratorCommand = (
  options: MockDataGenerationOptions
): string => {
  const outputPath = options.outputSubDir ?? ".";
  const locale = options.locale ? ` --locale ${options.locale}` : "";
  const maxArrayLength =
    options.maxArrayLength !== undefined
      ? ` --max-array-length ${options.maxArrayLength}`
      : "";
  return buildTypeSafeApiExecCommand(
    TypeSafeApiScript.GENERATE_MOCK_DATA,
    `--spec-path ${options.specPath} --output-path ${outputPath}${locale}${maxArrayLength}`
  );
};
