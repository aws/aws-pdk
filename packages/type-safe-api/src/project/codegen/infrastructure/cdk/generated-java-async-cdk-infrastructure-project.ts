/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import {
  GeneratedJavaCdkInfrastructureBaseProject,
  GeneratedJavaCdkInfrastructureBaseProjectOptions,
} from "./generated-java-cdk-infrastructure-base-project";
import {
  GenerationOptions,
  OtherGenerators,
  getHandlersProjectVendorExtensions,
} from "../../components/utils";

export interface GeneratedJavaAsyncCdkInfrastructureProjectOptions
  extends GeneratedJavaCdkInfrastructureBaseProjectOptions {}

export class GeneratedJavaAsyncCdkInfrastructureProject extends GeneratedJavaCdkInfrastructureBaseProject {
  constructor(options: GeneratedJavaAsyncCdkInfrastructureProjectOptions) {
    super(options);
  }

  public buildOpenApiGeneratorOptions(): GenerationOptions {
    return {
      generator: "java",
      specPath: this.options.specPath,
      generatorDirectory: OtherGenerators.JAVA_ASYNC_CDK_INFRASTRUCTURE,
      srcDir: this.srcDir,
      normalizers: {
        KEEP_ONLY_FIRST_TAG_IN_OPERATION: true,
      },
      extraVendorExtensions: {
        "x-infrastructure-package": this.packageName,
        "x-runtime-package": this.options.generatedJavaTypes.packageName,
        ...getHandlersProjectVendorExtensions(
          this,
          this.options.generatedHandlers
        ),
      },
      // Do not generate map/list types. Generator will use built in HashMap, ArrayList instead
      generateAliasAsModel: false,
    };
  }
}
