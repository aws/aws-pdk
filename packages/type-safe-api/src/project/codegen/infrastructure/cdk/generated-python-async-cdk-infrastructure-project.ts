/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import * as path from "path";
import {
  GeneratedPythonCdkInfrastructureBaseProject,
  GeneratedPythonCdkInfrastructureBaseProjectOptions,
} from "./generated-python-cdk-infrastructure-base-project";
import {
  GenerationOptions,
  OtherGenerators,
  getHandlersProjectVendorExtensions,
} from "../../components/utils";

export interface GeneratedPythonAsyncCdkInfrastructureProjectOptions
  extends GeneratedPythonCdkInfrastructureBaseProjectOptions {}

export class GeneratedPythonAsyncCdkInfrastructureProject extends GeneratedPythonCdkInfrastructureBaseProject {
  constructor(options: GeneratedPythonAsyncCdkInfrastructureProjectOptions) {
    super(options);
  }

  public buildOpenApiGeneratorOptions(): GenerationOptions {
    return {
      generator: "python-nextgen",
      specPath: this.options.specPath,
      generatorDirectory: OtherGenerators.PYTHON_ASYNC_CDK_INFRASTRUCTURE,
      // Tell the generator where python source files live
      srcDir: this.moduleName,
      normalizers: {
        KEEP_ONLY_FIRST_TAG_IN_OPERATION: true,
      },
      extraVendorExtensions: {
        "x-runtime-module-name": this.options.generatedPythonTypes.moduleName,
        // Spec path relative to the source directory
        "x-relative-spec-path": path.join("..", this.options.specPath),
        ...getHandlersProjectVendorExtensions(
          this,
          this.options.generatedHandlers
        ),
      },
    };
  }
}
