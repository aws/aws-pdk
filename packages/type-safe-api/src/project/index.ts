/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
export * from "./model/type-safe-api-model-build";
export * from "./model/type-safe-api-async-model-build";
export * from "./model/smithy/smithy-project-definition";
export * from "./model/openapi/open-api-project-definition";
export * from "./type-safe-api-project";
export * from "./model/type-safe-api-model-project";
export * from "./model/smithy/smithy-definition";
export * from "./model/openapi/open-api-definition";
export * from "./type-safe-websocket-api-project";
export * from "./model/type-safe-websocket-api-model-project";
export * from "./model/smithy/smithy-async-definition";
export * from "./model/openapi/open-api-async-definition";
export * from "./model/smithy/types";
export * from "./model/openapi/open-api-model-project";
export * from "./model/openapi/open-api-async-model-project";
export * from "./model/smithy/smithy-async-model-project";
export * from "./model/smithy/smithy-model-project";
export * from "./languages";
export * from "./types";
export * from "./typescript-project-options";
export * from "./python-project-options";
export * from "./java-project-options";
