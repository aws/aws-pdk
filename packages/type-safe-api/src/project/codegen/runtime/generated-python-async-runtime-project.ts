/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import {
  GeneratedPythonRuntimeBaseProject,
  GeneratedPythonRuntimeBaseProjectOptions,
} from "./generated-python-runtime-base-project";
import { GenerationOptions, OtherGenerators } from "../components/utils";

/**
 * Configuration for the generated python types project
 */
export interface GeneratedPythonAsyncRuntimeProjectOptions
  extends GeneratedPythonRuntimeBaseProjectOptions {}

/**
 * Python project containing types generated using OpenAPI Generator CLI
 */
export class GeneratedPythonAsyncRuntimeProject extends GeneratedPythonRuntimeBaseProject {
  constructor(options: GeneratedPythonAsyncRuntimeProjectOptions) {
    super(options);
  }

  protected buildOpenApiGeneratorOptions(): GenerationOptions {
    return {
      generator: "python",
      specPath: this.options.specPath,
      generatorDirectory: OtherGenerators.PYTHON_ASYNC_RUNTIME,
      additionalProperties: {
        packageName: this.moduleName,
        projectName: this.name,
      },
      // Tell the generator where python source files live
      srcDir: this.moduleName,
      normalizers: {
        KEEP_ONLY_FIRST_TAG_IN_OPERATION: true,
      },
    };
  }
}
