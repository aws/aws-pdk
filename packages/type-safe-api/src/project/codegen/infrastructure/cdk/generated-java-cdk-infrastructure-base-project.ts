/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import * as path from "path";
import { ProjectUtils } from "@aws/monorepo";
import { DependencyType, Task } from "projen";
import { JavaProject, JavaProjectOptions } from "projen/lib/java";
import {
  CodeGenerationSourceOptions,
  GeneratedWithOpenApiGeneratorOptions,
} from "../../../types";
import { TypeSafeApiCommandEnvironment } from "../../components/type-safe-api-command-environment";
import {
  buildCodegenCommandArgs,
  buildTypeSafeApiExecCommand,
  CodegenOptions,
  TypeSafeApiScript,
} from "../../components/utils";
import { GeneratedHandlersProjects } from "../../generate";
import { GeneratedJavaRuntimeBaseProject } from "../../runtime/generated-java-runtime-base-project";

export interface GeneratedJavaInfrastructureBaseOptions
  extends JavaProjectOptions,
    GeneratedWithOpenApiGeneratorOptions {}

export interface GeneratedJavaCdkInfrastructureBaseProjectOptions
  extends GeneratedJavaInfrastructureBaseOptions,
    CodeGenerationSourceOptions {
  /**
   * The generated java types
   */
  readonly generatedJavaTypes: GeneratedJavaRuntimeBaseProject;
  /**
   * Generated handlers projects
   */
  readonly generatedHandlers: GeneratedHandlersProjects;
}

export abstract class GeneratedJavaCdkInfrastructureBaseProject extends JavaProject {
  /**
   * Options configured for the project
   */
  protected readonly options: GeneratedJavaCdkInfrastructureBaseProjectOptions;

  /**
   * Source directory
   */
  protected readonly srcDir: string;

  /**
   * Java package name
   */
  protected readonly packageName: string;

  /**
   * Code generation task
   */
  protected readonly generateTask: Task;

  constructor(options: GeneratedJavaCdkInfrastructureBaseProjectOptions) {
    super({
      ...options,
      sample: false,
      junit: false,
    });
    TypeSafeApiCommandEnvironment.ensure(this);

    this.options = options;
    this.packageName = `${this.pom.groupId}.${this.name}.infra`;
    this.srcDir = path.join(
      "src",
      "main",
      "java",
      ...this.packageName.split(".")
    );

    [
      `software.aws/pdk@${ProjectUtils.getPdkVersion()}`,
      "software.constructs/constructs@10.3.0",
      "software.amazon.awscdk/aws-cdk-lib@2.133.0",
      "io.github.cdklabs/cdknag@2.28.60",
      "org.projectlombok/lombok@1.18.30",
      "com.fasterxml.jackson.core/jackson-databind@2.17.0",
      `io.github.cdklabs/projen@0.80.10`,
      `${options.generatedJavaTypes.pom.groupId}/${options.generatedJavaTypes.pom.artifactId}@${options.generatedJavaTypes.pom.version}`,
    ]
      .filter(
        (dep) =>
          !this.deps.tryGetDependency(dep.split("@")[0], DependencyType.RUNTIME)
      )
      .forEach((dep) => this.addDependency(dep));

    // Pin constructs version
    this.deps.removeDependency(
      "software.constructs/constructs",
      DependencyType.BUILD
    );
    this.addDependency("software.constructs/constructs@10.3.0");

    // Remove the projen test dependency since otherwise it takes precedence, causing projen to be unavailable at synth time
    this.deps.removeDependency("io.github.cdklabs/projen", DependencyType.TEST);

    // Add a dependency on the generated java types repository
    this.pom.addRepository({
      url: `file://${path.relative(
        this.outdir,
        options.generatedJavaTypes.outdir
      )}/dist/java`,
      id: `${options.generatedJavaTypes.pom.groupId}-${options.generatedJavaTypes.pom.artifactId}-repo`,
    });

    const generateTask = this.addTask("generate");
    this.generateTask = generateTask;
    generateTask.exec(
      buildTypeSafeApiExecCommand(
        TypeSafeApiScript.GENERATE_NEXT,
        this.buildGenerateCommandArgs()
      )
    );
    // Copy the parsed spec into the resources directory so that it's included in the jar
    generateTask.exec("mkdir -p src/main/resources");
    generateTask.exec(
      `cp -f ${this.options.specPath} src/main/resources/.api.json`
    );
    // Absolute path of this project is required for determining the path to the handlers jar,
    // since java executes from the jar which could be anywhere in the filesystem (eg the .m2 directory).
    // While ugly, since this is written at build time and not checked in it remains portable.
    generateTask.exec(
      "echo $(pwd) > src/main/resources/project-absolute-path.txt"
    );

    this.gitignore.addPatterns(".tsapi-manifest", ".openapi-generator");

    this.preCompileTask.spawn(generateTask);
  }

  public buildGenerateCommandArgs = () => {
    return buildCodegenCommandArgs(this.buildCodegenOptions());
  };

  protected abstract buildCodegenOptions(): CodegenOptions;
}
