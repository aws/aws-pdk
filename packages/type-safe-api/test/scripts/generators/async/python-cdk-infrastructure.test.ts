/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import os from "os";
import * as path from "path";
import { exec } from "projen/lib/util";
import { getTestHandlerProjects } from "./utils";
import { GeneratedPythonAsyncCdkInfrastructureProject } from "../../../../src/project/codegen/infrastructure/cdk/generated-python-async-cdk-infrastructure-project";
import { GeneratedPythonAsyncRuntimeProject } from "../../../../src/project/codegen/runtime/generated-python-async-runtime-project";
import { withTmpDirSnapshot } from "../../../project/snapshot-utils";

describe("Python Async Infrastructure Code Generation Script Unit Tests", () => {
  it.each(["single.yaml"])("Generates With %s", (spec) => {
    const specPath = path.resolve(
      __dirname,
      `../../../resources/specs/async/${spec}`
    );

    const snapshot = withTmpDirSnapshot(os.tmpdir(), (outdir) => {
      exec(`cp ${specPath} ${outdir}/spec.yaml`, {
        cwd: path.resolve(__dirname),
      });
      const clientOutdir = path.join(outdir, "client");
      const client = new GeneratedPythonAsyncRuntimeProject({
        name: "test-client",
        moduleName: "test_client",
        authorEmail: "me@example.com",
        authorName: "test",
        version: "1.0.0",
        outdir: clientOutdir,
        specPath: "../spec.yaml",
      });
      const infraOutdir = path.join(outdir, "infra");
      const project = new GeneratedPythonAsyncCdkInfrastructureProject({
        name: "test-infra",
        moduleName: "test_infra",
        authorEmail: "me@example.com",
        authorName: "test",
        version: "1.0.0",
        outdir: infraOutdir,
        specPath: "../spec.yaml",
        generatedPythonTypes: client,
        generatedHandlers: {},
      });
      project.synth();
      exec(
        `${path.resolve(
          __dirname,
          "../../../../scripts/type-safe-api/run.js generate"
        )} ${project.buildGenerateCommandArgs()}`,
        {
          cwd: infraOutdir,
        }
      );
    });

    expect(snapshot["infra/test_infra/api.py"]).toMatchSnapshot();
    expect(snapshot["infra/test_infra/__init__.py"]).toMatchSnapshot();
    expect(snapshot["infra/test_infra/mock_integrations.py"]).toMatchSnapshot();
  });

  it.each(["handlers.yaml", "inline-body.yaml"])(
    "Generates Functions for %s",
    (specFile) => {
      const specPath = path.resolve(
        __dirname,
        `../../../resources/specs/async/${specFile}`
      );

      const snapshot = withTmpDirSnapshot(os.tmpdir(), (outdir) => {
        exec(`cp ${specPath} ${outdir}/spec.yaml`, {
          cwd: path.resolve(__dirname),
        });
        const { runtimes, handlers } = getTestHandlerProjects(outdir);

        const infraOutdir = path.join(outdir, "infra");
        const project = new GeneratedPythonAsyncCdkInfrastructureProject({
          name: "test-infra",
          moduleName: "test_infra",
          authorEmail: "me@example.com",
          authorName: "test",
          version: "1.0.0",
          outdir: infraOutdir,
          specPath: "../spec.yaml",
          generatedPythonTypes: runtimes.python,
          generatedHandlers: handlers,
        });
        project.synth();
        exec(
          `${path.resolve(
            __dirname,
            "../../../../scripts/type-safe-api/run.js generate"
          )} ${project.buildGenerateCommandArgs()}`,
          {
            cwd: infraOutdir,
          }
        );
      });

      expect(snapshot["infra/test_infra/functions.py"]).toMatchSnapshot();
    }
  );
});
