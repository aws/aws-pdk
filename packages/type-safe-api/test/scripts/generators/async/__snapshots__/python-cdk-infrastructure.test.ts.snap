// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Python Async Infrastructure Code Generation Script Unit Tests Generates Functions for handlers.yaml 1`] = `
"from aws_cdk import Duration
from aws_cdk.aws_lambda import (
    Function, Runtime, Tracing, Code
)
from aws_pdk.type_safe_api import SnapStartFunction
from os import path
from pathlib import Path



class JavaOneFunction(SnapStartFunction):
    """
    Lambda function construct which points to the java implementation of JavaOne
    """
    def __init__(self, scope, id, **kwargs):
        super().__init__(scope, id,
            runtime=Runtime.JAVA_17,
            handler="test.test-java-handlers.handlers.JavaOneHandler",
            code=Code.from_asset(path.join(str(Path(__file__).absolute().parent), "..",
                "../java-handlers/dist/java/test/com.aws.pdk.test.handlers/1.0.0/com.aws.pdk.test.handlers-1.0.0.jar",
            )),
            tracing=Tracing.ACTIVE,
            timeout=Duration.seconds(30),
            **kwargs,
        )

class JavaTwoFunction(SnapStartFunction):
    """
    Lambda function construct which points to the java implementation of JavaTwo
    """
    def __init__(self, scope, id, **kwargs):
        super().__init__(scope, id,
            runtime=Runtime.JAVA_17,
            handler="test.test-java-handlers.handlers.JavaTwoHandler",
            code=Code.from_asset(path.join(str(Path(__file__).absolute().parent), "..",
                "../java-handlers/dist/java/test/com.aws.pdk.test.handlers/1.0.0/com.aws.pdk.test.handlers-1.0.0.jar",
            )),
            tracing=Tracing.ACTIVE,
            timeout=Duration.seconds(30),
            **kwargs,
        )

class PythonOneFunction(Function):
    """
    Lambda function construct which points to the python implementation of PythonOne
    """
    def __init__(self, scope, id, **kwargs):
        super().__init__(scope, id,
            runtime=Runtime.PYTHON_3_11,
            handler="test_python_handlers.python_one.handler",
            code=Code.from_asset(path.join(str(Path(__file__).absolute().parent), "..",
                "../python-handlers/dist/lambda",
            )),
            tracing=Tracing.ACTIVE,
            timeout=Duration.seconds(30),
            **kwargs,
        )

class PythonTwoFunction(Function):
    """
    Lambda function construct which points to the python implementation of PythonTwo
    """
    def __init__(self, scope, id, **kwargs):
        super().__init__(scope, id,
            runtime=Runtime.PYTHON_3_11,
            handler="test_python_handlers.python_two.handler",
            code=Code.from_asset(path.join(str(Path(__file__).absolute().parent), "..",
                "../python-handlers/dist/lambda",
            )),
            tracing=Tracing.ACTIVE,
            timeout=Duration.seconds(30),
            **kwargs,
        )

class TypescriptOneFunction(Function):
    """
    Lambda function construct which points to the typescript implementation of TypescriptOne
    """
    def __init__(self, scope, id, **kwargs):
        super().__init__(scope, id,
            runtime=Runtime.NODEJS_18_X,
            handler="index.handler",
            code=Code.from_asset(path.join(str(Path(__file__).absolute().parent), "..",
                "../typescript-handlers/dist/lambda",
                "typescript-one",
            )),
            tracing=Tracing.ACTIVE,
            timeout=Duration.seconds(30),
            **kwargs,
        )

class TypescriptTwoFunction(Function):
    """
    Lambda function construct which points to the typescript implementation of TypescriptTwo
    """
    def __init__(self, scope, id, **kwargs):
        super().__init__(scope, id,
            runtime=Runtime.NODEJS_18_X,
            handler="index.handler",
            code=Code.from_asset(path.join(str(Path(__file__).absolute().parent), "..",
                "../typescript-handlers/dist/lambda",
                "typescript-two",
            )),
            tracing=Tracing.ACTIVE,
            timeout=Duration.seconds(30),
            **kwargs,
        )"
`;

exports[`Python Async Infrastructure Code Generation Script Unit Tests Generates Functions for inline-body.yaml 1`] = `
"from aws_cdk import Duration
from aws_cdk.aws_lambda import (
    Function, Runtime, Tracing, Code
)
from aws_pdk.type_safe_api import SnapStartFunction
from os import path
from pathlib import Path



class JavaFunction(SnapStartFunction):
    """
    Lambda function construct which points to the java implementation of Java
    """
    def __init__(self, scope, id, **kwargs):
        super().__init__(scope, id,
            runtime=Runtime.JAVA_17,
            handler="test.test-java-handlers.handlers.JavaHandler",
            code=Code.from_asset(path.join(str(Path(__file__).absolute().parent), "..",
                "../java-handlers/dist/java/test/com.aws.pdk.test.handlers/1.0.0/com.aws.pdk.test.handlers-1.0.0.jar",
            )),
            tracing=Tracing.ACTIVE,
            timeout=Duration.seconds(30),
            **kwargs,
        )

class PythonFunction(Function):
    """
    Lambda function construct which points to the python implementation of Python
    """
    def __init__(self, scope, id, **kwargs):
        super().__init__(scope, id,
            runtime=Runtime.PYTHON_3_11,
            handler="test_python_handlers.python.handler",
            code=Code.from_asset(path.join(str(Path(__file__).absolute().parent), "..",
                "../python-handlers/dist/lambda",
            )),
            tracing=Tracing.ACTIVE,
            timeout=Duration.seconds(30),
            **kwargs,
        )

class TypeScriptFunction(Function):
    """
    Lambda function construct which points to the typescript implementation of TypeScript
    """
    def __init__(self, scope, id, **kwargs):
        super().__init__(scope, id,
            runtime=Runtime.NODEJS_18_X,
            handler="index.handler",
            code=Code.from_asset(path.join(str(Path(__file__).absolute().parent), "..",
                "../typescript-handlers/dist/lambda",
                "type-script",
            )),
            tracing=Tracing.ACTIVE,
            timeout=Duration.seconds(30),
            **kwargs,
        )"
`;

exports[`Python Async Infrastructure Code Generation Script Unit Tests Generates With single.yaml 1`] = `
"from dataclasses import fields
from aws_pdk.type_safe_api import TypeSafeRestApi, TypeSafeApiIntegration
from test_client.api.operation_config import OperationLookup, OperationConfig
from os import path
from pathlib import Path

SPEC_PATH = path.join(str(Path(__file__).absolute().parent), "../../spec.yaml")

class Api(TypeSafeRestApi):
    """
    Type-safe construct for the API Gateway resources defined by your model.
    This construct is generated and should not be modified.
    """
    def __init__(self, scope, id, integrations: OperationConfig[TypeSafeApiIntegration], **kwargs):
        super().__init__(scope, id,
            **kwargs,
            integrations={ field.name: getattr(integrations, field.name) for field in fields(integrations) },
            spec_path=SPEC_PATH,
            operation_lookup=OperationLookup,
        )
"
`;

exports[`Python Async Infrastructure Code Generation Script Unit Tests Generates With single.yaml 2`] = `"#"`;

exports[`Python Async Infrastructure Code Generation Script Unit Tests Generates With single.yaml 3`] = `undefined`;
